src/components/elements/table/GenericTable.tsx

```diff
@@ -298,41 +293,14 @@ const GenericTable = <T extends { id: string }>({
   );
   const hasHeaders = columns.find((c) => !!c.header);
 
-  // initially undefined so we can early return and avoid state flicker
-  const [selected, setSelected] = useState<string[]>();
-
-  const selectableRowIds = useMemo(() => {
-    if (!selectable) return [];
-    if (!isRowSelectable) return rows.map((r) => r.id);
-    return rows.filter(isRowSelectable).map((r) => r.id);
-  }, [rows, selectable, isRowSelectable]);
-
-  const handleSelectAllClick = useCallback(
-    (event: React.ChangeEvent<HTMLInputElement>) => {
-      if (event.target.checked) {
-        setSelected(selectableRowIds);
-      } else {
-        setSelected([]);
-      }
-    },
-    [selectableRowIds]
-  );
-
-  const handleSelectRow = useCallback(
-    (row: T) =>
-      setSelected((old) => {
-        if (!old) return undefined;
-        return old.includes(row.id) ? without(old, row.id) : [...old, row.id];
-      }),
-    []
-  );
-
-  // Clear selection when data changes
-  useEffect(() => setSelected([]), [rows]);
-
-  useEffect(() => {
-    if (selected) onChangeSelectedRowIds?.(selected);
-  }, [selected, onChangeSelectedRowIds]);
+  const { selected, selectableRowIds, handleSelectAllClick, handleSelectRow } =
+    useTableSelection({
+      selectable: !!selectable,
+      isRowSelectable,
+      rows,
+      selectedCtrl: selectedProp,
+      onChangeSelectedCtrl: onChangeSelectedRowIds,
```

Comment:
nit: `Ctrl` suffix feels off here because this callback is also used for components that use the selection in an uncontrolled manner. In that case they rely on this onChange handler to know which rows are currently selected